jerry-core/ecma/operations/ecma-promise-object.c

```diff
@@ -0,0 +1,711 @@
+  }
+
+  ecma_free_values_collection (reactions, false);
+} /* ecma_promise_trigger_reactions */
+
+/**
+ * Reject a Promise with a reason.
+ *
+ * See also: ES2015 25.4.1.7
+ */
+static void
+reject_promise (ecma_value_t promise, /**< promise */
+                ecma_value_t reason) /**< reason for reject */
+{
+  ecma_object_t *obj_p = ecma_get_object_from_value (promise);
+
+  JERRY_ASSERT (ecma_promise_get_state (obj_p) == ECMA_PROMISE_STATE_PENDING);
+
+  ecma_promise_set_result (obj_p, ecma_copy_value_if_not_object (reason));
+
+  ecma_promise_object_t *promise_p = (ecma_promise_object_t *) obj_p;
+  ecma_promise_trigger_reactions (promise_p->reject_reactions, reason);
+  promise_p->reject_reactions = ecma_new_values_collection (NULL, 0, false);
+  /* Free all fullfill_reactions. */
+  ecma_free_values_collection (promise_p->fulfill_reactions, false);
+  promise_p->fulfill_reactions = ecma_new_values_collection (NULL, 0, false);
+
+  ecma_promise_set_state (obj_p, ECMA_PROMISE_STATE_REJECTED);
+} /* reject_promise */
+
+/**
+ * Fulfill a Promise with a value.
+ *
+ * See also: ES2015 25.4.1.4
+ */
+static void
+fulfill_promise (ecma_value_t promise, /**< promise */
+                 ecma_value_t value) /**< fulfilled value */
+{
+  ecma_object_t *obj_p = ecma_get_object_from_value (promise);
+
+  JERRY_ASSERT (ecma_promise_get_state (obj_p) == ECMA_PROMISE_STATE_PENDING);
+
+  ecma_promise_set_result (obj_p, ecma_copy_value_if_not_object (value));
+
+  ecma_promise_object_t *promise_p = (ecma_promise_object_t *) obj_p;
+  ecma_promise_trigger_reactions (promise_p->fulfill_reactions, value);
+  promise_p->fulfill_reactions = ecma_new_values_collection (NULL, 0, false);
+  /* Free all reject_reactions. */
+  ecma_free_values_collection (promise_p->reject_reactions, false);
+  promise_p->reject_reactions = ecma_new_values_collection (NULL, 0, false);
+
+  ecma_promise_set_state (obj_p, ECMA_PROMISE_STATE_FULFILLED);
+} /* fulfill_promise */
+
+/**
+ * Native handler for Promise Reject Function.
+ *
+ * See also: ES2015 25.4.1.3.1
+ *
+ * @return ecma value of undefined.
+ */
+static ecma_value_t
+ecma_promise_reject_handler (const ecma_value_t function, /**< the function itself */
+                             const ecma_value_t this, /**< this_arg of the function */
+                             const ecma_value_t argv[], /**< argument list */
+                             const ecma_length_t argc) /**< argument number */
+{
+  JERRY_UNUSED (this);
+  JERRY_UNUSED (argc);
+
+  ecma_string_t *str_0 = ecma_new_ecma_string_from_uint32 (0);
+  ecma_string_t *str_1 = ecma_new_ecma_string_from_uint32 (1);
+  ecma_object_t *function_p = ecma_get_object_from_value (function);
+  /* 2. */
+  ecma_value_t promise = ecma_op_object_get (function_p, str_0);
+  /* 1. */
+  JERRY_ASSERT (ecma_is_promise (ecma_get_object_from_value (promise)));
+  /* 3. */
+  ecma_value_t already_resolved = ecma_op_object_get (function_p, str_1);
```

Comment:
I don't really understand this part. The function_p is a public function where we define a "1" property? What ensures that we can define this property?